src/ServiceControl/Operations/RetryConfirmationProcessor.cs

```diff
@@ -0,0 +1,83 @@
+﻿namespace ServiceControl.Operations
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Contracts.MessageFailures;
+    using Infrastructure.DomainEvents;
+    using MessageFailures;
+    using NServiceBus.Logging;
+    using NServiceBus.Transport;
+    using Raven.Abstractions.Commands;
+    using Raven.Abstractions.Data;
+    using Recoverability;
+
+    class RetryConfirmationProcessor
+    {
+        public const string SuccessfulRetryHeader = "ServiceControl.Retry.Successful";
+        const string RetryUniqueMessageIdHeader = "ServiceControl.Retry.UniqueMessageId";
+
+        public RetryConfirmationProcessor(IDomainEvents domainEvents)
+        {
+            this.domainEvents = domainEvents;
+        }
+
+        public (IReadOnlyList<MessageContext>, IReadOnlyCollection<ICommandData>) Process(List<MessageContext> contexts)
+        {
+            var storedContexts = new List<MessageContext>(contexts.Count);
+            var allCommands = new List<ICommandData>(contexts.Count);
+
+            foreach (var context in contexts)
+            {
+                try
+                {
+                    var commands = ProcessOne(context);
+                    allCommands.AddRange(commands);
+                    storedContexts.Add(context);
```

Comment:
why do we need to return a new collection that would contain the same elements as the `context` parameter?

Reply:
You mean the `storedContexts`? We need to return it because it may contain less elements than the input parameter if one of the messages failed to be processed. In that case there will be no commands associated with it and also we won't be forwarding that message to the error.log queue nor publishing integration events.

Reply:
but it doesn't really look like the returned elements is smaller than the incoming one for the retry confirmation processor? There isn't really anything in this part of the code path that is expected to go wrong?

Reply:
Ah yes. Sorryg, was thinking about failed messages.

Reply:
We discussed that and we concluded that we can remove the complexity for the retry confirmation path.